lib/puppet_x/puppetlabs/scheduled_task/trigger.rb

```diff
@@ -1,674 +1,678 @@
+      def self.time_valid?(time)
+        Time.parse("2016-5-1 #{time}")
+        true
+      rescue
+        false
       end
 
-      required = V2::EVENT_BASED_TRIGGER_MAP.value?(manifest_hash['schedule']) ? [] : %w{start_time}
+      # canonicalize given manifest hash
+      # throws errors if hash structure is invalid
+      # does not throw errors when invalid types are specified
+      # @returns original object with downcased keys
+      def self.canonicalize_and_validate(manifest_hash)
+        raise TypeError unless manifest_hash.is_a?(Hash)
+        manifest_hash = downcase_keys(manifest_hash)
 
-      required.each do |field|
-        next if manifest_hash.key?(field)
-        raise ArgumentError.new("Must specify '#{field}' when defining a trigger")
-      end
+        # check for valid key usage
+        invalid_keys = manifest_hash.keys - ValidKeys
+        raise ArgumentError, "Unknown trigger option(s): #{Puppet::Parameter.format_value_for_display(invalid_keys)}" unless invalid_keys.empty?
 
-      start_time_valid = begin Time.parse("2016-5-1 #{manifest_hash['start_time']}"); true rescue false; end
-      raise ArgumentError.new("Invalid start_time value: #{manifest_hash['start_time']}") unless start_time_valid
-      # The start_time must be canonicalized to match the format that the rest of the code expects
-      manifest_hash['start_time'] = format_time(Time.parse(manifest_hash['start_time'])) unless manifest_hash['start_time'].nil?
+        unless ValidScheduleKeys.include?(manifest_hash['schedule'])
+          raise ArgumentError, "Unknown schedule type: #{manifest_hash['schedule'].inspect}"
+        end
 
-      # specific setting rules for schedule types
-      case manifest_hash['schedule']
-      when 'monthly'
-        if manifest_hash.key?('on')
-          if manifest_hash.key?('day_of_week') || manifest_hash.key?('which_occurrence')
-            raise ArgumentError.new("Neither 'day_of_week' nor 'which_occurrence' can be specified when creating a monthly date-based trigger")
-          end
-        elsif manifest_hash.key?('which_occurrence') || manifest_hash.key?('day_of_week')
-          raise ArgumentError.new('which_occurrence cannot be specified as an array') if manifest_hash['which_occurrence'].is_a?(Array)
+        required = V2::EVENT_BASED_TRIGGER_MAP.value?(manifest_hash['schedule']) ? [] : ['start_time']
 
-          %w{day_of_week which_occurrence}.each do |field|
-            next if manifest_hash.key?(field)
-            raise ArgumentError.new("#{field} must be specified when creating a monthly day-of-week based trigger")
+        required.each do |field|
+          next if manifest_hash.key?(field)
+          raise ArgumentError, "Must specify '#{field}' when defining a trigger"
+        end
+
+        start_time_valid = time_valid?(manifest_hash['start_time'])
+        raise ArgumentError, "Invalid start_time value: #{manifest_hash['start_time']}" unless start_time_valid
```

Comment:
I don't understand why these errors don't need to be decorated like _("Invalid start_time value...")

Reply:
TBH, neither do I? But rubocop isn't yelling in this case.

![kinda confused](https://media.giphy.com/media/1oJLpejP9jEvWQlZj4/giphy.gif)